[FLINK-23832][docs] Update DataStream API Integration page (#16871)
apache/flink: docs/content/docs/dev/table/data_stream_api.md

types. It covers working with event-time and watermarks. It discusses how to declare a primary key and
changelog mode for the input and output streams.

The example above shows how the final result is computed incrementally by continuously emitting row-wise
updates for each incoming record. However, in cases where the input streams are finite (i.e. *bounded*),

> a result can be computed more efficiently by leveraging batch processing principles. Both DataStream

Reviewer: Perhaps a brief explanation of what this means or a link to more info?

Author: I added a concrete example and a link to the dynamic table section.

API and Table API offer a specialized *batch runtime mode*.

The following example illustrates that the unified pipeline is able to process both batch and streaming data
by just switching a flag.

{{< tabs "61a8a0b8-c38b-48f3-b52e-563546139380" >}}
{{< tab "Java" >}}
```java
import org.apache.flink.api.common.RuntimeExecutionMode;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;

// setup DataStream API
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

// set the batch runtime mode
env.setRuntimeMode(RuntimeExecutionMode.BATCH);

// uncomment this for streaming mode
// env.setRuntimeMode(RuntimeExecutionMode.STREAMING);

// setup Table API
// the table environment adopts the runtime mode during initialization
StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env);

// define the same pipeline as above

// prints in BATCH mode:
// +I[Bob, 10]
// +I[Alice, 112]

// prints in STREAMING mode:
// +I[Alice, 12]
// +I[Bob, 10]
// -U[Alice, 12]

> // +U[Alice, 112]

Reviewer: Wondering whether this is best left as code comments or whether it would be better to explain it in words

Author: I would leave it as a code comment here because this is only a page for integration with DataStream API. I don't want to cover all details about how changelog processing works. This should be a dedicated page that requires a lot of thought. Also what switching to batch mode means actually belongs on a different "concepts" page, but I briefly explained it here to have at least something.

env.setRuntimeMode(RuntimeExecutionMode.BATCH);
StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env);

// or

// set mode explicitly for StreamTableEnvironment
// it will be propagated to StreamExecutionEnvironment during planning
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env, EnvironmentSettings.inBatchMode());
```
{{< /tab >}}
{{< tab "Scala" >}}
```scala
import org.apache.flink.api.common.RuntimeExecutionMode
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.table.api.bridge.scala.StreamTableEnvironment
import org.apache.flink.table.api.EnvironmentSettings

// adopt mode from StreamExecutionEnvironment
val env = StreamExecutionEnvironment.getExecutionEnvironment
env.setRuntimeMode(RuntimeExecutionMode.BATCH)
val tableEnv = StreamTableEnvironment.create(env)

// or

// set mode explicitly for StreamTableEnvironment
val env = StreamExecutionEnvironment.getExecutionEnvironment
val tableEnv = StreamTableEnvironment.create(env, EnvironmentSettings.inBatchMode)
```
{{< /tab >}}
{{< /tabs >}}

> Setting the runtime mode has the following implications (among others):

Reviewer 2: nit: "Setting _this_ runtime mode" or "Setting the runtime mode [to batch]"?

- All sources must declare themselves as bounded. Otherwise an exception is thrown.

- Currently, table sources must emit insert-only changes. Otherwise an exception is thrown.

- Progressive watermarks are neither generated nor used in operators. However, sources emit a maximum
watermark before shutting down.

- Exchanges between tasks might be blocking according to the [`execution.batch-shuffle-mode`]({{< ref "docs/deployment/config" >}}#execution-batch-shuffle-mode).
This also means potentially less resource requirements compared to executing the same pipeline in streaming mode.

> - Checkpointing must be disabled. Artificial state backends are inserted.

Reviewer 2: Some of these points (bounded sources, insert-only sources, disabled checkpointing) aren't really implications but rather prerequisites. I think we should list those separately in a "the following needs to be true" kind of sense, and then list the rest as implications.

{{< top >}}

> Batch Runtime Mode

Reviewer: This whole section seems kind of general.  Do you think it should be in a separate section and then linked to?

Author: yes, it should be a dedicated page. but this is future work. we need a batch mode page that is independent of API without any DataStream API or Table API example.

- Operators need a sufficient amount of [off-heap memory]({{< ref "docs/deployment/memory/mem_setup_tm" >}}#managed-memory)
for sorting and other intermediate results.

- All table operations must be available in batch mode. Currently, some of them are only available in
streaming mode. Please check the corresponding Table API & SQL pages.

A batch execution has the following implications (among others):

- Progressive watermarks are neither generated nor used in operators. However, sources emit a maximum
watermark before shutting down.

- Exchanges between tasks might be blocking according to the [`execution.batch-shuffle-mode`]({{< ref "docs/deployment/config" >}}#execution-batch-shuffle-mode).
This also means potentially less resource requirements compared to executing the same pipeline in streaming mode.

- Checkpointing is disabled. Artificial state backends are inserted.

- Table operations don't produce incremental updates but only a complete final result which converts
to an insert-only changelog stream.

{{< hint info >}}
Since batch processing can be considered as *a special case of stream processing*, we recommend implementing

> a streaming pipeline first as it is the most general implementation for both bounded and unbounded data.

Reviewer: Should a blurb be added about how Flink runs more optimally this way?

{{< /hint >}}

The following example shows how to play around with batch mode using the [DataGen table source]({{< ref "docs/connectors/table/datagen" >}}). Many sources offer options that implicitly make the connector bounded, for example,

> by defining an terminating offset or timestamp. In our example, we limit the number of rows with the

Reviewer: ```suggestion
by defining a terminating offset or timestamp. In our example, we limit the number of rows with the
```

{{< hint info >}}
Since batch processing can be considered as *a special case of stream processing*, we recommend implementing
a streaming pipeline first as it is the most general implementation for both bounded and unbounded data.

In theory, a streaming pipeline can execute all operators. However, in practice, some operations might
not make much sense as they would lead to ever-growing state and are therefore not supported. A global
sort would be an example that is only available in batch mode. Simply put: it should be possible to

> run a working streaming pipeline in batch and but not necessarily vice versa.

Reviewer: ```suggestion
run a working streaming pipeline in batch mode but not necessarily vice versa.
```

`TypeInformation` is a `CompositeType`, it will be flattened in the first level when deriving a table's
schema.

{{< hint warning >}}

> Many times the DataStream API is unable to extract a more specific `TypeInformation` based on reflection.

Reviewer: ```suggestion
Many times the DataStream API is unable to extract a more specific `TypeInformation` based on [reflection](https://www.oracle.com/technical-resources/articles/java/javareflection.html).
```

schema.

{{< hint warning >}}

> Many times the DataStream API is unable to extract a more specific `TypeInformation` based on [reflection](https://www.oracle.com/technical-resources/articles/java/javareflection.html).

Reviewer: ```suggestion
The DataStream API is not always able to extract a more specific `TypeInformation` based on [reflection](https://www.oracle.com/technical-resources/articles/java/javareflection.html).
```
